Modify TableModel to handle absorption model from XSPEC table

jjlk opened the pull request:
Hi @cdeil and @joleroi , 
I added a class method to `TableModel` to import models from XSPEC (see [flx2tab](https://heasarc.gsfc.nasa.gov/ftools/caldb/help/flx2tab.html)).

Maybe we could propose this format on https://gamma-astro-data-formats.readthedocs.io/en/latest/? @cdeil, @registerrier , some volunteers?

I also made optional the logarithm scale value on the y-axis and I added a plot function because, as I understood it,  the SpectralModel `plot` method was really attached to have some physical units for the y-values.

Thanks!

cdeil commented:
Please add `$GAMMAPY_EXTRA/datasets/ebl/ebl_franceschini.fits.gz` so that tests run:
https://travis-ci.org/gammapy/gammapy/jobs/184942465#L997

You can directly add the file to `master` in `gammapy-extra`, unless you want it reviewed.
The usual things are: please add a README file what it is in the folder, and don't commit too large files without discussion (ask first if it's >> 1 MB).

Concerning format -- yes, I'd pretty much always would prefer formats are described over at `gamma-astro-data-formats`, instead of in the Gammapy docs. Especially for things like this, where the same data is of interest to several science tools.
I don't have time for this one. If you do, the start would be to either make a pull request with a very short and simple description of what you have here (as a starting point) or an issue proposing it, and CC a few people that might be able to give useful feedback.

cdeil commented on gammapy/spectrum/models.py:
```diff
@@ -574,30 +577,129 @@ class TableModel(SpectralModel):
+        filename = str(make_path(filename))
+
+        # Check if parameter value is in range
+        table_param = fits.open(filename)['PARAMETERS']
```
Can you try and use `astropy.table.Table` here?
We use it almost everywhere else instead of `astropy.io.fits.BinTable` directly.
To read a specific HDU:
```python
Table.read(filename, hdu='MY_HDU_NAME')
```

cdeil commented on gammapy/spectrum/models.py:
```diff
@@ -574,30 +577,129 @@ class TableModel(SpectralModel):
+        filename = '$GAMMAPY_EXTRA/datasets/ebl/ebl_franceschini.fits.gz'
+        table_model = TableModel.from_file(filename=filename, param=0.3)
+        """
+        energy_col = 'ENERG'
```
You're using this `energy_col = 'ENERG'` two times below, with string formatting.
IMO it would be simpler to remove it, and just use the column names directly below.

cdeil commented on gammapy/spectrum/models.py:
```diff
@@ -574,30 +577,129 @@ class TableModel(SpectralModel):
+
+        Examples
+        --------
+        Fill table from a model
```
To get proper syntax highlighting, you have to indent the code in the docstring by four spaces, like this:
```
Here's some code:

    print('hello!')
```

cdeil commented on gammapy/spectrum/models.py:
```diff
@@ -574,30 +577,129 @@ class TableModel(SpectralModel):
                                    kind='cubic')
 
+    @classmethod
+    def from_file(cls, filename, param):
```
What is `param`???

Mention it with one line (or add a `Parameters` section) in the docstring?

jjlk commented:
Hi @cdeil ,
> Please add $GAMMAPY_EXTRA/datasets/ebl/ebl_franceschini.fits.gz so that tests run:
https://travis-ci.org/gammapy/gammapy/jobs/184942465#L997

The files are already there. The test failed on Travis because of a precision problem. Locally it works fine for me. For now I just changed the parameter value for the test.

> Concerning format -- yes, I'd pretty much always would prefer formats are described over at gamma-astro-data-formats, instead of in the Gammapy docs. Especially for things like this, where the same data is of interest to several science tools.
I don't have time for this one. If you do, the start would be to either make a pull request with a very short and simple description of what you have here (as a starting point) or an issue proposing it, 
and CC a few people that might be able to give useful feedback.

Oké, I'll have a look at that later.

Thanks for your comments, I took all of them into account!

joleroi commented:
> I added a plot function because, as I understood it, the SpectralModel plot method was really attached to have some physical units for the y-values.

Doesn't it work if you pass ``flux_unit=''``?

I had a discussion with @adonath about this and we decided it might be better do rename ``flux_unit -> y_unit`` everywhere. Its just not top priority I think, but if you want to do it, go ahead (would also need to be done in ``gammapy/spectrum/results.py``)

joleroi commented on gammapy/spectrum/models.py:
```diff
@@ -574,30 +577,131 @@ class TableModel(SpectralModel):
-                                   kind='cubic')
+    @classmethod
+    def from_file(cls, filename, param):
+        """A Table containing aborbed values from a XSPEC model
```
typo: ``absorbed`` (?)

cdeil replied:
Maybe call the method `read_<something>` instead of `from_file`?
E.g. `read_xspec_absorbed` or `read_ebl`?

I think we always use the term "read" for classmethods reading from files, or at least I'm +1 to do that.

joleroi commented on gammapy/spectrum/models.py:
```diff
@@ -574,30 +577,131 @@ class TableModel(SpectralModel):
+        table_param = Table.read(filename, hdu='PARAMETERS')
+        param_min = table_param['MINIMUM']
+        param_max = table_param['MAXIMUM']
+        assert(param >= param_min and param <= param_max), "Parameter\
```
I've never seen a hard line break within a string literal coded like this, So I don't know if it's "allowed". What I've seen many time is
```
a = " Very long string, It's so long that we need a new "
      "line somewhere"
```

cdeil replied:
A coding rule for Gammapy (and Astropy / ... most projects): Never use `assert` to check inputs, raise an exception instead.


Here:
```python
if param_min <= param <= param_max:
    raise ValueError(<message here>)
```

joleroi commented on gammapy/spectrum/models.py:
```diff
@@ -574,30 +577,131 @@ class TableModel(SpectralModel):
+        energy = energy_bounds.log_centers
+
+        # Get spectrum values (no interpolation, take closest value for param)
+        table_spectra = Table.read(filename, hdu='SPECTRA')
```
Is it ok to hardcode the ``hdu`` names? Is this always the same for the format you're using, or is it better to pass the hdu names as kwargs?

cdeil commented on gammapy/spectrum/models.py:
```diff
@@ -574,30 +577,131 @@ class TableModel(SpectralModel):
+            File containing the XSPEC model
+        param : float
+            Model parameter value
+        Examples
```
I don't think this docstring will render correctly.
You need to put an empty line before the `Examples` heading, and I think also colon before the code block.

Can you please check locally with
```
python setup.py build_docs
open docs/_build/html/index.html
```
and then check this docstring on the HGPS page?

We will add automated checks for docstrings in the future, there's a discussion for Astropy generally how to best do it. For now, checking docstrings is done manually.

cdeil commented on gammapy/spectrum/tests/test_models.py:
```diff
@@ -159,3 +159,14 @@ def test_to_sherpa(spectrum):
+@requires_data('gammapy-extra')
+def test_table_model_from_file():
+    filename = '$GAMMAPY_EXTRA/datasets/ebl/ebl_franceschini.fits.gz'
+    energy_range = (0.03, 10)
```
No need for local variables here IMO, since only used once.
Just pass as keyword arguments below.

jjlk commented:
Hi @joleroi and @cdeil , 
> Doesn't it work if you pass flux_unit=''? 
> I had a discussion with @adonath about this and we decided it might be better do rename flux_unit -> y_unit everywhere. Its just not top priority I think, but if you want to do it, go ahead (would also need to be done in gammapy/spectrum/results.py)

I tried that and it didn't work. I didn't push a lot to investiguate where the problem was.

> Maybe call the method read_<something> instead of from_file?
> E.g. read_xspec_absorbed or read_ebl?

Done, `read_xspec_model`

> I don't think this docstring will render correctly.

Fixed

> A coding rule for Gammapy (and Astropy / ... most projects): Never use assert to check inputs, raise an exception instead.

Fixed

> Is it ok to hardcode the hdu names? Is this always the same for the format you're using, or is it better to pass the hdu names as kwargs?

Yes, it is a standard

> No need for local variables here IMO, since only used once.
> Just pass as keyword arguments below.

Fixed

And multiple typos fixed.

Thanks!
